From cce163404dcbf2328792ecada602c4c0c60c2bfc Mon Sep 17 00:00:00 2001
From: Sushanth Sowmyan <khorgath@gmail.com>
Date: Tue, 29 Mar 2016 11:21:23 -0700
Subject: [PATCH 0588/1164] CDH-40076 : Backport HIVE-12937 :
 DbNotificationListener unable to clean up old
 notification events (Sushanth Sowmyan, reviewed
 by Sergey Shelukhin)

(cherry picked from commit 1de97bc5fad323ae3bd48ebb39e6e68a3581e099)

Change-Id: Ie641e8c49137d3016e4acc04d3b42a5bdfaf13d4
---
 .../listener/TestDbNotificationListener.java       |   18 ++++++++++++++++++
 .../apache/hadoop/hive/metastore/ObjectStore.java  |    2 +-
 2 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/itests/hcatalog-unit/src/test/java/org/apache/hive/hcatalog/listener/TestDbNotificationListener.java b/itests/hcatalog-unit/src/test/java/org/apache/hive/hcatalog/listener/TestDbNotificationListener.java
index 91cc03e..d7a71a2 100644
--- a/itests/hcatalog-unit/src/test/java/org/apache/hive/hcatalog/listener/TestDbNotificationListener.java
+++ b/itests/hcatalog-unit/src/test/java/org/apache/hive/hcatalog/listener/TestDbNotificationListener.java
@@ -57,6 +57,7 @@
 public class TestDbNotificationListener {
 
   private static final Log LOG = LogFactory.getLog(TestDbNotificationListener.class.getName());
+  private static final int EVENTS_TTL = 30;
   private static Map<String, String> emptyParameters = new HashMap<String, String>();
   private static IMetaStoreClient msClient;
   private static Driver driver;
@@ -68,6 +69,7 @@ public static void connectToMetastore() throws Exception {
     HiveConf conf = new HiveConf();
     conf.setVar(HiveConf.ConfVars.METASTORE_EVENT_LISTENERS,
         DbNotificationListener.class.getName());
+    conf.setVar(HiveConf.ConfVars.METASTORE_EVENT_DB_LISTENER_TTL, String.valueOf(EVENTS_TTL)+"s");
     conf.setBoolVar(HiveConf.ConfVars.HIVE_SUPPORT_CONCURRENCY, false);
     conf.setBoolVar(HiveConf.ConfVars.FIRE_EVENTS_FOR_DML, true);
     conf.setVar(HiveConf.ConfVars.DYNAMICPARTITIONINGMODE, "nonstrict");
@@ -565,4 +567,20 @@ public void sqlInsertPartition() throws Exception {
     assertEquals(firstEventId + 19, event.getEventId());
     assertEquals(HCatConstants.HCAT_DROP_PARTITION_EVENT, event.getEventType());
   }
+
+  @Test
+  public void cleanupNotifs() throws Exception {
+    Database db = new Database("cleanup1","no description","file:/tmp", emptyParameters);
+    msClient.createDatabase(db);
+    msClient.dropDatabase("cleanup1");
+
+    NotificationEventResponse rsp = msClient.getNextNotification(firstEventId, 0, null);
+    assertEquals(2, rsp.getEventsSize());
+
+    // sleep for expiry time, and then fetch again
+    Thread.sleep(EVENTS_TTL * 2 * 1000); // sleep twice the TTL interval - things should have been cleaned by then.
+
+    NotificationEventResponse rsp2 = msClient.getNextNotification(firstEventId, 0, null);
+    assertEquals(0, rsp2.getEventsSize());
+  }
 }
\ No newline at end of file
diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java b/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java
index b4a9f96..77bcc60 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java
@@ -7571,7 +7571,7 @@ public void cleanNotificationEvents(int olderThan) {
       query.declareParameters("java.lang.Integer tooOld");
       Collection<MNotificationLog> toBeRemoved = (Collection) query.execute(tooOld);
       if (toBeRemoved != null && toBeRemoved.size() > 0) {
-        pm.deletePersistent(toBeRemoved);
+        pm.deletePersistentAll(toBeRemoved);
       }
       commited = commitTransaction();
     } finally {
-- 
1.7.9.5

